### PR TITLE
SpaceAfterComma rule

### PR DESCRIPTION
**Pull request by msoedov**
#146 That's a really straightforward implementation. However there are 39 errors! for this rule in the project. Some of those are really easy to fix but there are couple places where spaces after comma used for some kind of visual indent like

```elixir
  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_),     do: ["lib"]
```

What should we do with such corner cases?

### CODE REVIEW

**Comment by lpil** [comment]
I think that we should enforce >= 1 spaces by default, with configurable modes for = 1 spaces, and =0 spaces.

File: lib/dogma/rule/space_after_comma.ex [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    |> Enum.map(fn line ->  error(rule, line) end)
+  end
+
+  defp check_line(rule, {_, line}) do
+    @spaces_pattern
+    |> Regex.scan(line)
+    |> Enum.any?(fn [_, group] ->
+      String.length(group) != rule.spaces
+      end)
+  end
```
Comment (lpil) [inline lib/dogma/rule/space_after_comma.ex]: Can we indent this like this please?

```elixir
defp check_line(rule, {_, line}) do
  @spaces_pattern
  |> Regex.scan(line)
  |> Enum.any?(fn [_, group] ->
    String.length(group) != rule.spaces
  end)
end
```

File: lib/dogma/rule/space_after_comma.ex [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  defp error(rule, {pos, _}) do
+    %Error{
+      rule:     __MODULE__,
+      message:  "Should be #{rule.spaces} spaces after comma",
+      line: Dogma.Script.line(pos),
```
Comment (lpil) [inline lib/dogma/rule/space_after_comma.ex]: Indent this value to match the others please.

**Comment by msoedov** [comment]
@lpil that's a good point, I'll updated it today

File: lib/dogma/rule/space_after_comma.ex [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  @moduledoc """
+  A rule that enforces to use `spaces` after comma.
+
+  This rule can be configured with the `spaces` option, which allows you to
+  specify your own number of spaces after comma.
+  """
+  @spaces_pattern ~r/\,(\s*).+/
+
+  def test(rule, script) do
+    script.lines
```
Comment (lpil) [inline lib/dogma/rule/space_after_comma.ex]: This will throw errors for commas inside strings.

File: test/dogma/rule/space_after_comma_test.exs [hunk 4]
```diff
@@ -0,0 +1,62 @@
+defmodule Dogma.Rule.SpaceAfterCommaTest do
+  use RuleCase, for: SpaceAfterComma
+
+  defp lint(script, rule) do
+    script |> Script.parse!( "" ) |> fn s -> Rule.test(rule, s) end.()
```
Comment (lpil) [inline test/dogma/rule/space_after_comma_test.exs]: Can we use the testing style used in the other rule tests please?

File: test/dogma/rule/space_after_comma_test.exs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    assert [] == errors
+  end
+
+  defp error_on_line(line, spaces) do
+    %Error{
+      line: Dogma.Script.line(line),
+      message: "Should be #{spaces} spaces after comma",
+      rule: SpaceAfterComma
+    }
+  end
```
Comment (lpil) [inline test/dogma/rule/space_after_comma_test.exs]: Add a test for commas inside strings.

**Comment by lpil** [comment]
This is looking good, thanks! :)

File: lib/dogma/rule/space_after_comma.ex [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+  This rule can be configured with the `spaces` option, which allows you to
+  specify your own number of spaces after comma.
+  """
+  @spaces_pattern ~r/\,(\s*).+/
+  @string_pattern ~r/\"(.*?,.*?)\"/
+
+  def test(rule, script) do
+    script.lines
+    |> Enum.map(&drop_strings/1)
```
Comment (msoedov) [inline lib/dogma/rule/space_after_comma.ex]: Dunno if it possible to ignore string in a better way.

**Comment by msoedov** [comment]
Current version also checks module docks and comments which is kinda side effect but I think should be fine.

File: lib/dogma/rule/space_after_comma.ex (hunk 6 above)
Comment (rrrene) [inline lib/dogma/rule/space_after_comma.ex]: I think @lpil wanted you to use the `processed_source` instead of the `source` of the Script. It is set via [sigils](https://github.com/lpil/dogma/blob/master/lib/dogma/util/script_sigils.ex) and [strings](https://github.com/lpil/dogma/blob/master/lib/dogma/util/script_strings.ex) helpers and [should not contain any sigils and strings](https://github.com/lpil/dogma/blob/410d88070a511e872d340e88f879a15385962a3e/lib/dogma/script.ex#L89) :+1:

Comment (lpil) [inline lib/dogma/rule/space_after_comma.ex]: That's the one, sorry I wasn't clear. Thanks Rene! :)

**Comment by msoedov** [comment]
Any other thoughts? Is there are any other corner cases?

**Comment by lpil** [comment]
Fantastic work! Thank you so much :)

**Comment by lpil** [comment]
Nice looking code too

**Comment by msoedov** [comment]
> Nice looking code too

I never feel right when I'm using regexp but thank you!

**Comment by lpil** [comment]
Right tool for the job I think!
